java/arcs/android/storage/service/BindingContext.kt

```diff
@@ -190,7 +217,47 @@ class BindingContext(
         }
     }
 
+    /**
+     * Increment an active message count and send it to the active messages count flow.
+     *
+     * Calls to [idle] will wait for this value to reach 0 before checking store activity.
+     *
+     * Avoid using this for calls to [idle] itself, so as to not increment the activity counter.
+     */
+    private fun launchNonIdleAction(action: suspend () -> Unit) {
+        activeMessagesCountFlow.increment()
+        // TODO(b/168724138) - Switch to a channel/flow queue-based approach.
+        actionQueue
+            .launch(action)
+            .invokeOnCompletion { activeMessagesCountFlow.decrement() }
+    }
+
     companion object {
         private val nextId = atomic(0)
     }
 }
+
+@ExperimentalCoroutinesApi
+/**
+ * A simple helper class to provide serialized operation execution on the provided
+ * [CoroutineScope].
+ *
+ * TODO(b/168724138) - Switch to a channel/flow queue-based approach and remove this.
+ */
+class MutexOperationQueue(
+    private val scope: CoroutineScope
+) {
+    // A mutex to serialize the incoming actions.
+    // Since coroutine mutexes process lock-waiters in order, this services as a quick-and-dirty
+    // queue for us.
+    private val actionMutex = Mutex()
+
+    fun launch(action: suspend () -> Unit): Job {
+        // [CoroutineStart.UNDISPATCHED] here to ensures that the actions are started in order.
+        return scope.launch(start = CoroutineStart.UNDISPATCHED) {
```

Comment:
Note:
CoroutineStart.UNDISPATCHED + mutex cannot ensure the observed order when the service api calls are remote i.e. binder #1 held the mutex and executing the action at its (binder) thread context due to CoroutineStart.UNDISPATCHED. binder #2 (arrived first) and binder #3 (arrived later) are waiting for the mutex. When the mutex is released, either #2 or #3 is resumed in the `scope`'s coroutine context.

flow-ish solution should work.

Reply:
According to docs for `Mutex.lock`, suspended callers are released in FIFO order.

Reply:
ah good to know, thanks!